[SPARK-34953][CORE][SQL] Add the code change for adding the DateType in the infer schema while reading in CSV and JSON

SaurabhChawla100 opened the pull request:
### What changes were proposed in this pull request?
Till now there is no support in infer schema for the DateType format while reading the CSV/json. In this PR, code change is done to support the DateType when inferschema set true in the options

### Why are the changes needed?
Many times there are multiple columns which are DateType but after inferred from schema they are added as StringType in the schema and than there is need to convert into to_date before we start the running any query. After this change DateType will be added in the schema instead of the StringType

### Does this PR introduce _any_ user-facing change?
No


### How was this patch tested?
Unit test added , also testing is done while running read command on spark shell.

SaurabhChawla100 committed:
Add the code change for adding the DateType in the infer schema

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchemaSuite.scala:
```diff
@@ -192,4 +192,13 @@ class CSVInferSchemaSuite extends SparkFunSuite with SQLHelper {
   }
+
+  test("SPARK-34953 - DateType should be inferred when user defined format are provided") {
+    val options = new CSVOptions(Map("dateFormat" -> "dd-MM-yyyy",
```
does it work with default date pattern?

SaurabhChawla100 replied:
Yes this works with default date patter which is 'yyyy-MM-dd', If we don't specify in the option dateFormat option it will check with the default value of dateFormat.

`2,23232,hello,2016-11-11`

```
val p =  spark.read.format("csv").option("header", "false").option("delimiter", ",").option("inferSchema", "true").load("/testDir/test.csv")

p.schema
res0: org.apache.spark.sql.types.StructType = StructType(StructField(_c0,IntegerType,true), StructField(_c1,IntegerType,true), StructField(_c2,StringType,true), StructField(_c3,DateType,true))
```

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala:
```diff
@@ -168,6 +169,15 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
+    }
+  }
+
+  private def tryParseDateFormat(field: String): DataType = {
```
Logically you should try smaller data one first.

SaurabhChawla100 replied:
done

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala:
```diff
@@ -168,6 +169,15 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
+  }
+
+  private def tryParseDateFormat(field: String): DataType = {
+    if ((allCatch opt new SimpleDateFormat(
```
Should avoid using `SimpleDateFormat` from JDK7. It has multiple flaws such as disallowing exact match. You should use `DateFormatter`.

SaurabhChawla100 replied:
@HyukjinKwon  - Sorry , I forget to push the latest commit  where I was setting setLenient(false) , This will help with the exact match.

```
  val dtFormat = new SimpleDateFormat(options.dateFormat, Locale.US)
      dtFormat.setLenient(false)
      dtFormat.parse(field)}).isDefined) {
      DateType
    }
```

HyukjinKwon commented:
See also https://github.com/apache/spark/pull/23202. I think this JIRA is a duplicate of SPARK-26248

SaurabhChawla100 committed:
Add the code change for adding the DateType in the infer schema

SaurabhChawla100 commented:
> See also #23202. I think this JIRA is a duplicate of [SPARK-26248](https://issues.apache.org/jira/browse/SPARK-26248)

@HyukjinKwon - Yes it looks like the duplicate of SPARK-26248 , which was reverted.

SaurabhChawla100 commented:
@HyukjinKwon  - As per the points mentioned in the reverted PR, Would like to get some understanding

```
Problem 1.

#23202 (comment) - I left some examples there.

If there are multiple rows, and the first row is inferred as date type in the same partition,
It will not be able to infer timestamp afterward.
```
Should this not be the same problem with timestamp format also if we specify  like this 
```
2,23232,hello,2016-11-11
2,23232,hello1,2016.11.11
```
```

 val p =  spark.read.format("csv").option("header", "false").option("delimiter", ",").option("timestampFormat", "yyyy-MM-dd").option("inferSchema", "true").load("/testDir/test.csv")

p.schema
res0: org.apache.spark.sql.types.StructType = StructType(StructField(_c0,IntegerType,true), StructField(_c1,IntegerType,true), StructField(_c2,StringType,true), StructField(_c3,StringType,true))
```

Here for timestampFormat we are getting the StringDataType which is the last one proffered. I believe in real world scenario it is the case of data-corruption,  if we are directly saving that data into the file, instead of applying some filtering criteria which prevent multiple type of data format in the same column 


```
Problem 2.

#23202 (comment)

If legacy is on, we have ambiguity about date/timestamp pattern matching, because they can be arbitrarily set by users.
It does not do the exact match, which means it's not going to distinguish yyyy-MM and yyyy-MM-dd for input, for instane, 2010-10-10.

We are able to do this only when spark.sql.legacy.timeParser.enabled is disabled (by default), however, I was thinking it's going to introduce complexity.
I was thinking we could do it later when we remove spark.sql.legacy.timeParser.enabled. Date type inference isn't super important IMHO becase we infer timestamps.
I would like to talk about this further if anyone thinks differently. If the change isn't complicated then I thought, it should also be okay to go ahead.
```

Not Able to understand this point clearly, But was thinking if for dateFormatType validate if we can set dtFormat.setLenient(false) it will be able to distinguish the both ```yyyy-MM and yyyy-MM-dd ```

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:
```diff
@@ -131,6 +133,13 @@ private[sql] class JsonInferSchema(options: JSONOptions) extends Serializable {
             (allCatch opt timestampFormatter.parse(field)).isDefined) {
           TimestampType
+        } else if (options.allowDateFormat &&
+          (allCatch opt { val dtFormat = new SimpleDateFormat(
```
Can we avoid using `SimpleDateFormat`, and trying date type (smaller type) first before timestamp type?

SaurabhChawla100 replied:
done

SaurabhChawla100 committed:
Used the DateFormatter instead of java.text.SimpleDateFormat

SaurabhChawla100 commented:
@HyukjinKwon - I made the changes as per the review comments. Can you please check this PR.

HyukjinKwon commented:
ok to test

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala:
```diff
@@ -68,6 +68,8 @@ private[sql] class JSONOptions(
   val allowBackslashEscapingAnyCharacter =
     parameters.get("allowBackslashEscapingAnyCharacter").map(_.toBoolean).getOrElse(false)
+  val allowDateFormat =
+    parameters.get("allowDateFormat").map(_.toBoolean).getOrElse(false)
```
Shall we add this option into CSV too? I would name it like `inferDateType`. BTW, I am just curious. Why did you disable it specifically here in JSON?

HyukjinKwon replied:
We would also have to add the docs in https://github.com/apache/spark/blob/master/docs/sql-data-sources-json.md and https://github.com/apache/spark/blob/master/docs/sql-data-sources-csv.md

SaurabhChawla100 replied:
**Shall we add this option into CSV too? I would name it like inferDateType.** - Added for CSV also and change the name to inferDateType

 **BTW, I am just curious. Why did you disable it specifically here in JSON?** - This is to keep in sync with older version of spark, If someone wants to use the dateType, they can enable it in the option. This is just to prevent any migration issue to spark-3.2.0 from the older version.

SaurabhChawla100 replied:
**We would also have to add the docs in https://github.com/apache/spark/blob/master/docs/sql-data-sources-json.md and https://github.com/apache/spark/blob/master/docs/sql-data-sources-csv.md** - Added the doc for the option in the sql-data-sources-json.md but not able to find place to add the option in sql-data-sources-csv.md

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala:
```diff
@@ -38,6 +39,13 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
+    options.zoneId,
+    options.locale,
+    legacyFormat = SIMPLE_DATE_FORMAT,
+    isParsing = true)
```
cc @MaxGekk FYI

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala:
```diff
@@ -38,6 +39,13 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
+    options.dateFormat,
+    options.zoneId,
+    options.locale,
+    legacyFormat = SIMPLE_DATE_FORMAT,
```
Let's use `FAST_DATE_FORMAT` for consistency with TimestampType

SaurabhChawla100 replied:
done

SparkQA commented:
**[Test build #139089 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139089/testReport)** for PR 32558 at commit [`29a5e43`](https://github.com/apache/spark/commit/29a5e4331c4fa5555301a484e52206174663f97a).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43610/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43610/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43610/

SparkQA commented:
**[Test build #139089 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139089/testReport)** for PR 32558 at commit [`29a5e43`](https://github.com/apache/spark/commit/29a5e4331c4fa5555301a484e52206174663f97a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139089/

SaurabhChawla100 committed:
Add the code change as per the review comments

SparkQA commented:
**[Test build #139098 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139098/testReport)** for PR 32558 at commit [`c743739`](https://github.com/apache/spark/commit/c743739c344cff0121dc3058611a27d0b6476dcd).

HyukjinKwon commented on docs/sql-data-sources-json.md:
```diff
@@ -155,6 +155,12 @@ Data source options of JSON can be set via:
   </tr>
+  <tr>
+    <td><code>inferDateType</code></td>
+    <td>None</td>
```
Let's change the default value to `false`. This was a mistake from https://github.com/apache/spark/pull/32204, and @itholic is working on it to fix. BTW, seems like CSV option is missing here.

SaurabhChawla100 replied:
done , added for csv also

HyukjinKwon commented:
I think we should also add the parameter at `csv` and `json` at `DataFrameWriter`, `DataFrameReader`, `DataStreamReader`, `DataStreamWriter` (and also `readwriter.py` and `streaming.py`). @itholic can you help review this one please?

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43619/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43619/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43619/

SaurabhChawla100 commented:
> I think we should also add the parameter at `csv` and `json` at `DataFrameWriter`, `DataFrameReader`, `DataStreamReader`, `DataStreamWriter` (and also `readwriter.py` and `streaming.py`). @itholic can you help review this one please?

Not able to understand what paramater is needed to add in the csv and json at at `DataFrameWriter`, `DataFrameReader`, `DataStreamReader`, `DataStreamWriter` (and also `readwriter.py` and `streaming.py`).

There is already `private var extraOptions = CaseInsensitiveMap[String](Map.empty)` and this code inserts the options that are added to it in 

```
def option(key: String, value: String): DataFrameReader = {
    this.extraOptions = this.extraOptions + (key -> value)
    this
  }
```

So this is working on giving in the option.

```
scala> spark.read.option("inferSchema", "true").option("inferDateType", "true").option("dateFormat", "yyyy-MM-dd").json(Seq("""{"a": {"b": 1, "c": "2021-02-26"}}""").toDS()).schema
res18: org.apache.spark.sql.types.StructType = StructType(StructField(a,StructType(StructField(b,LongType,true), StructField(c,DateType,true)),true))
```

SparkQA commented:
**[Test build #139113 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139113/testReport)** for PR 32558 at commit [`361ed49`](https://github.com/apache/spark/commit/361ed49bfd538a97b3223c1514364402e809ef9e).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43633/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43633/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43633/

SparkQA commented:
**[Test build #139098 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139098/testReport)** for PR 32558 at commit [`c743739`](https://github.com/apache/spark/commit/c743739c344cff0121dc3058611a27d0b6476dcd).
 * This patch **fails from timeout after a configured wait of `500m`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139098/

SparkQA commented:
**[Test build #139113 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139113/testReport)** for PR 32558 at commit [`361ed49`](https://github.com/apache/spark/commit/361ed49bfd538a97b3223c1514364402e809ef9e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139113/

HyukjinKwon commented:
Oh I meant this: https://github.com/apache/spark/blob/master/python/pyspark/sql/readwriter.py#L342-L350
These options are listed up as a parameter in Python side specifically. For CSV documentation, it's merged at https://github.com/apache/spark/pull/32658 so you could add the option in that page.

SparkQA commented:
**[Test build #139148 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139148/testReport)** for PR 32558 at commit [`3536100`](https://github.com/apache/spark/commit/3536100c4d84b01aea0c559f958f8b91525211cf).

SaurabhChawla100 committed:
Add the inferDateType as param in csv

SaurabhChawla100 commented:
> Oh I meant this: https://github.com/apache/spark/blob/master/python/pyspark/sql/readwriter.py#L342-L350
> These options are listed up as a parameter in Python side specifically. For CSV documentation, it's merged at #32658 so you could add the option in that page.

Added the inferDateType in the parameter , also added in the doc in sql-data-sources-csv.md

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43668/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43668/

SparkQA commented:
**[Test build #139151 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139151/testReport)** for PR 32558 at commit [`bb204ed`](https://github.com/apache/spark/commit/bb204ed31656568445d980a0daafc9e9bfdf368b).

SparkQA commented:
**[Test build #139151 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139151/testReport)** for PR 32558 at commit [`bb204ed`](https://github.com/apache/spark/commit/bb204ed31656568445d980a0daafc9e9bfdf368b).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139151/

SaurabhChawla100 committed:
fix style issue in python

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43671/

SparkQA commented:
**[Test build #139153 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139153/testReport)** for PR 32558 at commit [`3bd241a`](https://github.com/apache/spark/commit/3bd241abd28bef7943d80ee1072375e5933f66f2).

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43671/

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43673/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43673/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43671/

SparkQA commented:
**[Test build #139148 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139148/testReport)** for PR 32558 at commit [`3536100`](https://github.com/apache/spark/commit/3536100c4d84b01aea0c559f958f8b91525211cf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139148/

SparkQA commented:
**[Test build #139153 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139153/testReport)** for PR 32558 at commit [`3bd241a`](https://github.com/apache/spark/commit/3bd241abd28bef7943d80ee1072375e5933f66f2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139153/

SaurabhChawla100 committed:
add inferDatetype in streaming.py

SparkQA commented:
**[Test build #139169 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139169/testReport)** for PR 32558 at commit [`55b7f6c`](https://github.com/apache/spark/commit/55b7f6c3496f1498a1663547052d4c162454af2c).

SparkQA commented:
**[Test build #139169 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139169/testReport)** for PR 32558 at commit [`55b7f6c`](https://github.com/apache/spark/commit/55b7f6c3496f1498a1663547052d4c162454af2c).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43689/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139169/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43689/

SaurabhChawla100 committed:
resolving the conflict after merging with master

SparkQA commented:
**[Test build #139172 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139172/testReport)** for PR 32558 at commit [`3d46a3c`](https://github.com/apache/spark/commit/3d46a3c1c891648192acc7631b023df9dfd0d434).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43692/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43692/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43692/

SparkQA commented:
**[Test build #139172 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139172/testReport)** for PR 32558 at commit [`3d46a3c`](https://github.com/apache/spark/commit/3d46a3c1c891648192acc7631b023df9dfd0d434).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139172/

HyukjinKwon commented:
Looks making sense to me.

HyukjinKwon reviewed (approved):
It would be great if we can have a look from @MaxGekk though.

HyukjinKwon commented on docs/sql-data-sources-csv.md:
```diff
@@ -248,5 +248,11 @@ Data source options of CSV can be set via:
+  <tr>
+    <td><code>inferDateType</code></td>
+    <td>false</td>
+    <td>Infers all DateType format for the csv. If this is not set, it uses the default value, <code>false</code>.</td>
```
```suggestion
    <td>Infers all DateType format for the CSV. If this is not set, it uses the default value, <code>false</code>.</td>
```

HyukjinKwon commented on docs/sql-data-sources-json.md:
```diff
@@ -155,6 +155,12 @@ Data source options of JSON can be set via:
+  <tr>
+    <td><code>inferDateType</code></td>
+    <td>false</td>
+    <td>Infers all DateType format for the json. If this is not set, it uses the default value, <code>false</code>.</td>
```
```suggestion
    <td>Infers all DateType format for the JSON. If this is not set, it uses the default value, <code>false</code>.</td>
```

SaurabhChawla100 committed:
update the doc as per the comments suggestion

SaurabhChawla100 commented:
@HyukjinKwon  - Thank you for reviewing this PR.

@MaxGekk - Please take a look into this PR.

SparkQA commented:
**[Test build #139204 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139204/testReport)** for PR 32558 at commit [`78a7356`](https://github.com/apache/spark/commit/78a73560bb883d48b5022b4a2d5fde87cf10feb5).

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139204/

MaxGekk commented on docs/sql-data-sources-csv.md:
```diff
@@ -248,5 +248,11 @@ Data source options of CSV can be set via:
+  <tr>
+    <td><code>inferDateType</code></td>
+    <td>false</td>
+    <td>Infers all DateType format for the CSV. If this is not set, it uses the default value, <code>false</code>.</td>
```
What do you mean by `all DateType format`? I think we should say about the `dateFormat` option here.

MaxGekk commented on docs/sql-data-sources-json.md:
```diff
@@ -155,6 +155,12 @@ Data source options of JSON can be set via:
+  <tr>
+    <td><code>inferDateType</code></td>
+    <td>false</td>
+    <td>Infers all DateType format for the JSON. If this is not set, it uses the default value, <code>false</code>.</td>
```
**the** JSON ? Is it needed?

SaurabhChawla100 replied:
this is to specify for the JSON in this doc

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala:
```diff
@@ -109,6 +116,7 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
         case LongType => tryParseLong(field)
         case _: DecimalType => tryParseDecimal(field)
         case DoubleType => tryParseDouble(field)
+        case DateType => tryParseDateFormat(field)
```
Just curious why you try to infer dates before timestamps?

SaurabhChawla100 replied:
This is as per the suggestion in one of the review comments to use

https://github.com/apache/spark/pull/32558#discussion_r633020781

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala:
```diff
@@ -160,6 +168,16 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
+
+  private def tryParseDateFormat(field: String): DataType = {
+    if (options.inferDateType
+      && !dateFormatter.isInstanceOf[LegacySimpleDateFormatter]
```
I didn't get the check. Could you explain, please, why do you avoid the formatter?

SaurabhChawla100 replied:
It has to be LegacyFastDateFormatter, missed to changed it. Previously I was using the SimpleDateFormatter so added this LegacySimpleDateFormatter, Now since we are using the FastDateFormatter it has to be LegacyFastDateFormatter, Making that change.

If legacy is on, we have ambiguity about Datetype pattern matching, because they can be arbitrarily set by users.
It does not do the exact match, which means it's not going to distinguish yyyy-MM and yyyy-MM-dd for input, for instance, 2010-10-10.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVOptions.scala:
```diff
@@ -206,6 +206,12 @@ class CSVOptions(
   }
 
+  /**
+   * option to infer date Type in the schema
```
For me as an user, it is not clear the relation between the `inferSchema` option and this one. Does this option enable inferring independently from inferSchema`?

SaurabhChawla100 replied:
This option inferDateType is added to -> This is to keep in sync with older version of spark, If someone wants to use the dateType, they can enable it in the option. This is just to prevent any migration issue to spark-3.2.0 from the older version. if they don't enable this option inferDateType, It will infer it as StringType.

Where as on other hand inferSchema  is to enable the inferring of schema. 
If inferSchema is enabled and inferDateType option is enabled in that case on reading the schema is this will infer at data type as DateType format instead of  StringType

MaxGekk commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchemaSuite.scala:
```diff
@@ -192,4 +192,27 @@ class CSVInferSchemaSuite extends SparkFunSuite with SQLHelper {
     Seq("ko-KR", "ru-RU", "de-DE").foreach(checkDecimalInfer(_, DecimalType(7, 0)))
   }
+
+  test("SPARK-34953 - DateType should be inferred when user defined format are provided") {
```
The common convention is: `SPARK-XXXXX: ...`
```suggestion
  test("SPARK-34953: DateType should be inferred when user defined format are provided") {
```

SaurabhChawla100 replied:
done.

MaxGekk commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JsonInferSchemaSuite.scala:
```diff
@@ -112,4 +112,14 @@ class JsonInferSchemaSuite extends SparkFunSuite with SQLHelper {
     checkType(Map("inferTimestamp" -> "false"), json, StringType)
   }
+
+  test("SPARK-34953 - Allow DateType format while inferring") {
```
```suggestion
  test("SPARK-34953: Allow DateType format while inferring") {
```

SaurabhChawla100 replied:
done

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43727/

SaurabhChawla100 committed:
add the changes as per the review comments

SparkQA commented:
**[Test build #139207 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139207/testReport)** for PR 32558 at commit [`2689fd1`](https://github.com/apache/spark/commit/2689fd16ea5a6b0ae4b323fbbe1fada21f18bd5e).

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43727/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43727/

SaurabhChawla100 committed:
update LegacyFastDateFormatter instead of LegacySimpleDateFormatter

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139212/

SparkQA commented:
**[Test build #139213 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139213/testReport)** for PR 32558 at commit [`c93b873`](https://github.com/apache/spark/commit/c93b873bc5707d239784dec3fc20344cd0b1feb4).

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43730/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43730/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43735/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43736/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43735/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43735/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43736/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43736/

SparkQA commented:
**[Test build #139207 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139207/testReport)** for PR 32558 at commit [`2689fd1`](https://github.com/apache/spark/commit/2689fd16ea5a6b0ae4b323fbbe1fada21f18bd5e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139207/

SparkQA commented:
**[Test build #139213 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139213/testReport)** for PR 32558 at commit [`c93b873`](https://github.com/apache/spark/commit/c93b873bc5707d239784dec3fc20344cd0b1feb4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139213/

SaurabhChawla100 commented:
@MaxGekk - Thanks for reviewing the PR . I made the changes requested in the review comments and also replied to your comments. Can you please take look to the PR.

SaurabhChawla100 commented:
@MaxGekk , @HyukjinKwon - Thank you for reviewing this PR. Are we planning to move further with this PR or do we need more changes on this.  Please share your thoughts on this .

HyukjinKwon commented:
I would defer to @MaxGekk

nate-kuhl commented:
also @MaxGekk  ^^^

HyukjinKwon commented:
It's been added at https://github.com/apache/spark/pull/37327 and https://github.com/apache/spark/pull/23202
